### ISSUE TITLE
Status Code 500 on DirectMethod call after running for hours

### ISSUE THREAD

**Issue by ChrRing** [Azure/iotedge#711]
## Expected Behavior
Direct Method calls - triggered locally from a notebook, a VM, from the cloud as an Azure Functions App - should run for hours, days, weeks,... without changing the behaviour on a scheduled interval (here, every 15 seconds). The DirectMethod in this case does nothing else but checking whether an URL is accesible or not.

## Current Behavior
After some time (i.e., hours) with the DirectMethod call returning status code 200 (equals 0) for a succesful method call

## Steps to Reproduce

1. Implement custom Python module based on [this](https://docs.microsoft.com/en-us/azure/iot-edge/tutorial-python-module) documentation. Additional implement a function that checks the URL.
```
def device_method_callback(method_name, payload, user_context):
    if method_name == "directMethodTimerTrigger":
        utc_timestamp = datetime.datetime.utcnow().replace(tzinfo=datetime.timezone.utc).isoformat()
        msg = json.loads(payload)
        code = urllib.request.urlopen(msg['url']).getcode()
        print("URL called")
        output = {}
        output['ts'] = utc_timestamp
        output['url'] = msg
        output['code'] = code
        output = json.dumps(output)
        print("Data arrived")

    device_method_return_value = DeviceMethodReturnValue()
    device_method_return_value.response = output
    device_method_return_value.status = 200
    #To verify whether this works, also write directly to BLOB for the beginning; shall be replaced by a separate route to IoT Hub
    #accountName = '***'
    #accountKey = '***'
    #containerName = "testapp"
    #fileName = utc_timestamp + '.json'

    #blobService = BlockBlobService(account_name=accountName, account_key=accountKey)

    #blobService.create_blob_from_text(containerName, fileName, output)
    return device_method_return_value
```
2. Create a function that invokes the DirectMethod.
```
import datetime
import logging
import json
import logging
import azure.functions as func
from builtins import input
import iothub_service_client
from iothub_service_client import IoTHubDeviceMethod, IoTHubError

# The service connection string to authenticate with your IoT hub.
# Using the Azure CLI:
# az iot hub show-connection-string --hub-name {your iot hub name}
CONNECTION_STRING = "***"
DEVICE_ID = "***"
MODULE_ID = "directMethodUrlModule"

# Details of the direct method to call.
METHOD_NAME = "directMethodTimerTrigger"
METHOD_PAYLOAD = "{'url':'https://www.stackoverflow.com/'}"
TIMEOUT = 10
WAIT_COUNT = 10

def iothub_devicemethod_timertrigger():
    try:
        utc_timestamp = datetime.datetime.utcnow().replace(tzinfo=datetime.timezone.utc).isoformat()
        # Connect to your hub.
        iothub_device_method = IoTHubDeviceMethod(CONNECTION_STRING)
        # Call the direct method.
        response = iothub_device_method.invoke(DEVICE_ID, MODULE_ID, METHOD_NAME, METHOD_PAYLOAD, TIMEOUT)
        logging.info('Python timer trigger function ran at %s', utc_timestamp)
        logging.info( "" )
        logging.info( "Method called" )
        logging.info( "Method name       : {0}".format(METHOD_NAME) )
        logging.info( "Method payload    : {0}".format(METHOD_PAYLOAD) )
        logging.info( "" )
        logging.info( "Response status   : {0}".format(response.status) )
        logging.info( "Response payload  : {0}".format(response.payload) )

    except IoTHubError as iothub_error:
        logging.info( "" )
        logging.info( "Unexpected error {0}".format(iothub_error) )
        return

    except KeyboardInterrupt:
        logging.info( "" )
        logging.info( "IoTHubDeviceMethod sample stopped" )

def main(mytimer: func.TimerRequest) -> None:
    utc_timestamp = datetime.datetime.utcnow().replace(
        tzinfo=datetime.timezone.utc).isoformat()

    if mytimer.past_due:
        logging.info('The timer is past due!')

    iothub_devicemethod_timertrigger()
    
if __name__ == '__main__':
    print ( "Starting the IoT Hub Service Client DeviceManagement Python sample..." )
    print ( "    Connection string = {0}".format(CONNECTION_STRING) )
    print ( "    Device ID         = {0}".format(DEVICE_ID) )

    iothub_devicemethod_timertrigger()
```
3. Wait.

## Context (Environment)

> PRETTY_NAME="Raspbian GNU/Linux 9 (stretch)"
> NAME="Raspbian GNU/Linux"
> VERSION_ID="9"
> VERSION="9 (stretch)"
> ID=raspbian
> ID_LIKE=debian

### Architecture
arm32

### Container Operating System
Linux

#### iotedged
`iotedge 1.0.5 (d76e0316c6f324345d77c48a83ce836d09392699)`

#### Docker

> **Client:**
 Version:           3.0.2
 API version:       1.40
 Go version:        go1.11.1
 Git commit:        e042b58f
 Built:             Tue Dec  4 17:16:17 2018
 OS/Arch:           linux/arm
 Experimental:      false
**Server:**
 Engine:
  Version:          3.0.2
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.11.2
  Git commit:       74baf62
  Built:            Tue Dec  4 23:30:17 2018
  OS/Arch:          linux/arm
  Experimental:     false

## Logs
`iotedge logs directMethodUrlModule -f` returns:
```
Python 3.5.3 (default, Sep 27 2018, 17:25:39)
[GCC 6.3.0 20170516]

IoT Hub Client for Python
Starting the IoT Hub Python sample using protocol MQTT...
The sample is now waiting for messages and will indefinitely.  Press Ctrl-C to exit.
Error: Time:Fri Jan 11 11:29:39 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:wait_for_connection Line:385 Failure: connect failure 111.
Error: Time:Fri Jan 11 11:29:39 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:769 wait_for_connection failed
Error: Time:Fri Jan 11 11:29:39 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Fri Jan 11 11:29:39 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Fri Jan 11 11:29:39 2019 File:/usr/sdk/src/c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1000 Error: io_open failed
Error: Time:Fri Jan 11 11:29:39 2019 File:/usr/sdk/src/c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2122 failure connecting to address raspberrypi.
Error: Time:Fri Jan 11 11:29:40 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:wait_for_connection Line:385 Failure: connect failure 111.
Error: Time:Fri Jan 11 11:29:40 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:769 wait_for_connection failed
Error: Time:Fri Jan 11 11:29:40 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Fri Jan 11 11:29:40 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Fri Jan 11 11:29:40 2019 File:/usr/sdk/src/c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1000 Error: io_open failed
Error: Time:Fri Jan 11 11:29:40 2019 File:/usr/sdk/src/c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2122 failure connecting to address raspberrypi.
Error: Time:Fri Jan 11 11:29:42 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:wait_for_connection Line:385 Failure: connect failure 111.
Error: Time:Fri Jan 11 11:29:42 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:769 wait_for_connection failed
Error: Time:Fri Jan 11 11:29:42 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Fri Jan 11 11:29:42 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Fri Jan 11 11:29:42 2019 File:/usr/sdk/src/c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1000 Error: io_open failed
Error: Time:Fri Jan 11 11:29:42 2019 File:/usr/sdk/src/c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2122 failure connecting to address raspberrypi.
Error: Time:Fri Jan 11 11:29:46 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:wait_for_connection Line:385 Failure: connect failure 111.
Error: Time:Fri Jan 11 11:29:46 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:769 wait_for_connection failed
Error: Time:Fri Jan 11 11:29:46 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Fri Jan 11 11:29:46 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Fri Jan 11 11:29:46 2019 File:/usr/sdk/src/c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1000 Error: io_open failed
Error: Time:Fri Jan 11 11:29:46 2019 File:/usr/sdk/src/c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2122 failure connecting to address raspberrypi.
Error: Time:Fri Jan 11 11:29:54 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:wait_for_connection Line:385 Failure: connect failure 111.
Error: Time:Fri Jan 11 11:29:54 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:769 wait_for_connection failed
Error: Time:Fri Jan 11 11:29:54 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Fri Jan 11 11:29:54 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Fri Jan 11 11:29:54 2019 File:/usr/sdk/src/c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1000 Error: io_open failed
Error: Time:Fri Jan 11 11:29:54 2019 File:/usr/sdk/src/c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2122 failure connecting to address raspberrypi.
Error: Time:Fri Jan 11 11:30:10 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:wait_for_connection Line:385 Failure: connect failure 111.
Error: Time:Fri Jan 11 11:30:10 2019 File:/usr/sdk/src/c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:769 wait_for_connection failed
Error: Time:Fri Jan 11 11:30:10 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_open_complete Line:760 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Fri Jan 11 11:30:10 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:tlsio_openssl_open Line:1258 Failed opening the underlying I/O.
Error: Time:Fri Jan 11 11:30:10 2019 File:/usr/sdk/src/c/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:1000 Error: io_open failed
Error: Time:Fri Jan 11 11:30:10 2019 File:/usr/sdk/src/c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:2122 failure connecting to address raspberrypi.
```

## Additional Information

A lot of the code above is from different resources I don't have original source at hand. Credits for the code go to those that have initially developed it. Thanks!

**Comment by myagley** [comment]
This may be resolved by #685 which hasn't been released just yet. There is a deadlock in the dotnet runtime (which the Edge Hub uses) that we believe has been resolved by moving to version 2.1.6. We have been running some long haul tests to verify this.

Unfortunately, the only known workaround at this point is to restart the Edge Hub.

**Comment by bhall7718** [comment]
@myagley Was this rolled into the 1.0.6 changes, or tied up in 1.0.7.

**Comment by myagley** [comment]
The deadlock fix is in `1.0.6`. Note that this is only an issue on ARM32 devices.

The connectivity fixes for direct methods in general are hopefully coming in `1.0.7`. The fix is in https://github.com/Azure/azure-iot-sdk-csharp/pull/795 of the csharp SDK.

**Comment by lt72** [comment]
Closing this issue as it is fixed. Please re-open as needed.
